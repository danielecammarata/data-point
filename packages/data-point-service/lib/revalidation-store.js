const debug = require("debug")("data-point-service:cache");
const throttle = require("lodash/throttle");

/**
 * Maximum number of entries to store
 * NOTE: This value is only meant to add a cap to the keys we store. If there
 * is any strong feeling on making this part smarter such as creating a FIFO
 * sort of logic PRs are welcomed
 */
const MAX_STORE_SIZE = 10000;

/**
 * Time to wait for cleanup to trigger in case it's called more than once within
 * the range placed.
 */
const THROTTLE_WAIT = 1000; // milliseconds

/**
 * It marks and deletes all the keys that have their ttl expired
 * @param {Map} store Map Object that stores all the cached keys
 * @returns {Number} amount of entries deleted
 */
function clear(store) {
  const forDeletion = [];
  const now = Date.now();

  store.forEach((entry, key) => {
    // mark for deletion entries that have timed out
    if (now - entry.created > entry.ttl) {
      forDeletion.push(key);
    }
  });

  const forDeletionLength = forDeletion.length;
  // if nothing to clean then exit
  if (forDeletionLength === 0) {
    return 0;
  }

  debug(`local revalidation flags that timed out: ${forDeletion}`);
  forDeletion.forEach(key => store.delete(key));
  return forDeletionLength;
}

/**
 * @param {Map} store Map Object that stores all the cached keys
 * @param {Number} maxStoreSize Size to limit the store, if above this number keys will no longer be saved
 * @param {String} key key value
 * @param {Number} ttl time the key will be alive, expressed in milliseconds
 * @returns {Boolean} true if added, false if otherwise
 */
function add(store, maxStoreSize, key, ttl) {
  // do not add more than 10000 keys
  if (store.size > maxStoreSize) return false;
  store.set(key, { created: Date.now(), ttl });
  return true;
}

/**
 * @param {Map} store Map Object that stores all the cached keys
 * @param {String} key key value
 * @returns {Boolean} true, yes always true
 */
function remove(store, key) {
  store.delete(key);
  return true;
}

/**
 * True if a key exists, and it has not expired
 * @param {Map} store Map Object that stores all the cached keys
 * @param {String} key key value
 * @returns {Boolean} true if key exists, false otherwise
 */
function exists(store, key) {
  const entry = store.get(key);
  // checks entry exists and it has not timed-out
  return !!entry && Date.now() - entry.created < entry.ttl;
}

/**
 * @typedef {Object} RevalidationStore
 * @property {Map} store internal store instance
 * @property {function} add (key:String, ttl:Number) add entry
 * @property {function} remove (key:String) remove entry
 * @property {function} exists (key:String) check if entry exists
 * @property {function} clear () clears any key that has ttl expired, this method is throttled
 */

/**
 * Creates an in-memory Revalidation Controller
 * @returns {RevalidationStore}  Object
 */
function create() {
  const store = new Map();

  return {
    store,
    add: add.bind(null, store, MAX_STORE_SIZE),
    remove: remove.bind(null, store),
    exists: exists.bind(null, store),
    clear: throttle(clear.bind(null, store), THROTTLE_WAIT)
  };
}

module.exports = {
  MAX_STORE_SIZE,
  THROTTLE_WAIT,
  clear,
  add,
  remove,
  exists,
  create
};
