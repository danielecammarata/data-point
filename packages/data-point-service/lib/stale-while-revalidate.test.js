/* eslint-env jest */
const _ = require("lodash");

const mockDebug = jest.fn();
jest.mock("debug", () => {
  return () => mockDebug;
});

const mockLocalAdd = jest.fn();
const mockLocalRemove = jest.fn();

require("./revalidation-store");

jest.mock("./revalidation-store", () => {
  return {
    create: () => ({
      add: mockLocalAdd,
      remove: mockLocalRemove,
      clear: () => {}
    })
  };
});

const RedisController = require("./redis-controller");

const StaleWhileRevalidate = require("./stale-while-revalidate");

beforeEach(() => {
  jest.resetModules();
  delete require.cache[require.resolve("./stale-while-revalidate")];
});

describe("revalidationExternalFactory", () => {
  describe("add - add remote revalidation flag", () => {
    it("should call setSWRControlEntry", () => {
      const mockSetSWRControlEntry = jest
        .spyOn(RedisController, "setSWRControlEntry")
        .mockImplementation(() => true);

      const external = StaleWhileRevalidate.revalidationExternalFactory(
        "service"
      );

      expect(external.add("entryKey", "ttl")).toEqual(true);
      expect(mockSetSWRControlEntry).toBeCalledWith(
        "service",
        "entryKey",
        "ttl",
        StaleWhileRevalidate.SWR_CONTROL_REVALIDATING
      );
    });
  });

  describe("remove - remove remote revalidation flag", () => {
    it("should call setSWRControlEntry", () => {
      const mockDeleteSWRControlEntry = jest
        .spyOn(RedisController, "deleteSWRControlEntry")
        .mockImplementation(() => true);

      const external = StaleWhileRevalidate.revalidationExternalFactory(
        "service"
      );

      expect(external.remove("entryKey", "ttl")).toEqual(true);
      expect(mockDeleteSWRControlEntry).toBeCalledWith("service", "entryKey");
    });
  });

  describe("exists - check if remote revalidation flag exists", () => {
    it("should return true if value is not undefined", () => {
      const mockGetSWRControlEntry = jest
        .spyOn(RedisController, "getSWRControlEntry")
        .mockImplementation(() => Promise.resolve("foo"));

      const external = StaleWhileRevalidate.revalidationExternalFactory(
        "service"
      );

      return external.exists("entryKey").then(result => {
        expect(result).toEqual(true);
        expect(mockGetSWRControlEntry).toBeCalledWith("service", "entryKey");
      });
    });

    it("should return false if value is undefined", () => {
      jest
        .spyOn(RedisController, "getSWRControlEntry")
        .mockImplementation(() => Promise.resolve(undefined));

      const external = StaleWhileRevalidate.revalidationExternalFactory(
        "service"
      );

      return external.exists("entryKey").then(result => {
        expect(result).toEqual(false);
      });
    });
  });
});

describe("addEntry", () => {
  it("should add an entry to redis", () => {
    const mockSetSWRStaleEntry = jest
      .spyOn(RedisController, "setSWRStaleEntry")
      .mockImplementation(() => Promise.resolve(true));

    const mockSetSWRControlEntry = jest
      .spyOn(RedisController, "setSWRControlEntry")
      .mockImplementation(() => Promise.resolve(true));

    const cache = {
      ttl: "10s",
      staleWhileRevalidateTtl: "20s"
    };

    return StaleWhileRevalidate.addEntry(
      "service",
      "entryKey",
      "value",
      cache
    ).then(() => {
      expect(mockSetSWRStaleEntry).toBeCalledWith(
        "service",
        "entryKey",
        "value",
        cache.staleWhileRevalidateTtl
      );
      expect(mockSetSWRControlEntry).toBeCalledWith(
        "service",
        "entryKey",
        cache.ttl,
        StaleWhileRevalidate.SWR_CONTROL_STALE
      );
    });
  });
});

describe("getEntry", () => {
  it("should add an entry to redis", () => {
    const getSWRStaleEntry = jest
      .spyOn(RedisController, "getSWRStaleEntry")
      .mockImplementation(() => Promise.resolve(true));

    return StaleWhileRevalidate.getEntry("service", "entryKey", "value").then(
      () => {
        expect(getSWRStaleEntry).toBeCalledWith("service", "entryKey");
      }
    );
  });
});

describe("addRevalidationFlags", () => {
  it("should add local and external flags", () => {
    const revalidation = {};
    _.set(revalidation, "local.add", jest.fn());
    _.set(revalidation, "external.add", jest.fn(() => Promise.resolve(true)));

    return StaleWhileRevalidate.addRevalidationFlags(
      revalidation,
      "entryKey",
      "10s"
    ).then(() => {
      expect(revalidation.local.add).toBeCalledWith("entryKey", "10s");
      expect(revalidation.external.add).toBeCalledWith("entryKey", "10s");
    });
  });
});

describe("clearAllRevalidationFlags", () => {
  it("should clear local and external flags", () => {
    const revalidation = {};
    _.set(revalidation, "local.remove", jest.fn());
    _.set(
      revalidation,
      "external.remove",
      jest.fn(() => Promise.resolve(true))
    );

    return StaleWhileRevalidate.clearAllRevalidationFlags(
      revalidation,
      "entryKey"
    ).then(() => {
      expect(revalidation.local.remove).toBeCalledWith("entryKey");
      expect(revalidation.external.remove).toBeCalledWith("entryKey");
    });
  });
});

describe("getRevalidationState", () => {
  it("should get local and external flag states", () => {
    const revalidation = {};
    _.set(revalidation, "local.exists", jest.fn(() => "localState"));
    _.set(
      revalidation,
      "external.exists",
      jest.fn(() => Promise.resolve(false))
    );

    return StaleWhileRevalidate.getRevalidationState(
      revalidation,
      "entryKey"
    ).then(result => {
      expect(result.hasExternalEntryExpired).toEqual(true);
      expect(result.isRevalidatingLocally).toBeInstanceOf(Function);
      expect(result.isRevalidatingLocally()).toEqual("localState");
    });
  });

  it("should return hasExternalEntryExpired === true if exists == false", () => {
    const revalidation = {};
    _.set(revalidation, "local.exists", jest.fn(() => "localState"));
    _.set(
      revalidation,
      "external.exists",
      jest.fn(() => Promise.resolve(true))
    );

    return StaleWhileRevalidate.getRevalidationState(
      revalidation,
      "entryKey"
    ).then(result => {
      expect(result.hasExternalEntryExpired).toEqual(false);
      expect(result.isRevalidatingLocally).toBeInstanceOf(Function);
      expect(result.isRevalidatingLocally()).toEqual("localState");
    });
  });
});

describe("createRevalidationManager", () => {
  it("should create revalidation controller", () => {
    const revalidation = StaleWhileRevalidate.createRevalidationManager();

    expect(revalidation).toHaveProperty("local.add");
    expect(revalidation).toHaveProperty("external.add");
  });
});

describe("create", () => {
  it("should create stale while revalidate controller", () => {
    const swr = StaleWhileRevalidate.create();
    expect(swr).toMatchSnapshot();
  });
});
