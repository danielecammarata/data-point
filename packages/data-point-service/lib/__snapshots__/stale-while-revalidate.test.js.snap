// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create should create stale while revalidate controller 1`] = `
Object {
  "addEntry": [Function],
  "addRevalidationFlags": [Function],
  "clearAllRevalidationFlags": [Function],
  "getEntry": [Function],
  "getRevalidationState": [Function],
  "invalidateLocalFlags": [Function],
  "removeLocalRevalidationFlag": [MockFunction],
}
`;
